DataViews: Allow searching over array fieldtypes

dsas opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- Link this PR to its associated issue with an appropriate keyword: Closes, See, Follow up to, etc. -->
Closes #70782

<!-- In a few words, what is the PR actually doing? -->

Fixing a bug with searching over DataView fields that have the array field types.

Also updating the DataViews storybook to demonstrate that it works.

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->

This fixes a bug with searching over DataView fields that have the array fieldtype. Previously they would crash with _TypeError: input.trim is not a function_

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->

Treats every value as an array before normalising and searching over every value in that array.

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a post or page. -->
<!-- 2. Insert a heading block. -->
<!-- 3. etc. -->

I've added some tests, to test manually:

1. Open the story book
2. Go to DataViews
3. Search for "planet"
4. Note that it now returns rows that have "planet" as a category, not just in the description

### Testing Instructions for Keyboard
<!-- How can you test the changes by using the keyboard only? Please note, this is required for PRs that change the user interface (UI). This ensures the PR can be tested for any possible accessibility regressions. -->

dsas committed:
Dataviews: Allow searching over array fields

Searching over array fieldtypes with `enableGlobalSearch: true` threw a
TypeError because the `normalizeSearchInput` function expected a string
but received an array.

This change fixes that by treating every value as an array and adapting
the logic to match, before entering `normalizeSearchInput`.

dsas committed:
Dataviews: Add Category search to storybook

This shows off that the search works over array fields too.

oandregal commented on packages/dataviews/src/test/filter-and-sort-data-view.js:
```diff
@@ -71,6 +71,54 @@ describe( 'filters', () => {
+		// Should find items with "Space" in categories (case-insensitive)
+		// Plus the item with title "Space" (from title search which is already enabled)
+		expect( result ).toHaveLength( 10 );
+		expect( result.map( ( r ) => r.title ) ).toContain( 'Apollo' );
```
The logic sounds good, but the tests need updating. I've just landed an update to the test data this morning to improve it.

dsas committed:
Update tests to use new fixtures

Also changed the second test so it's not just returning all results so
it's clear that it is actually searching.

oandregal commented on packages/dataviews/src/test/filter-and-sort-data-view.js:
```diff
@@ -71,6 +71,54 @@ describe( 'filters', () => {
+		const { data: result } = filterSortAndPaginate(
+			data,
+			{
+				search: 'Moon',
```
`title` and `description` also have global search enabled, so this can give false positives because the 3 items (Europa, Io, Moon) have "moon" in the title or description. Perhaps change the search to `Satellite`?

oandregal reviewed (approved):
Thanks for the fix.

I'm pre-approving in the interest of velocity, though I think it'd be better to update the tests search string ([see](https://github.com/WordPress/gutenberg/pull/70785#discussion_r2224554563)).

dsas committed:
Switch test to use "satellite" so we can be sure it is not picking up on title / description

oandregal commented:
Ah, I forgot to mention something: this needs a changelog update :)
